src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -845,29 +834,25 @@ RunLocalExecution(CitusScanState *scanState, DistributedExecution *execution)
 														execution->defaultTupleDest,
 														isUtilityCommand);
 
-	/*
-	 * We're deliberately not setting execution->rowsProcessed here. The main reason
-	 * is that modifications to reference tables would end-up setting it both here
-	 * and in AdaptiveExecutor. Instead, we set executorState here and skip updating it
-	 * for reference table modifications in AdaptiveExecutor.
-	 */
-	EState *executorState = ScanStateGetExecutorState(scanState);
-	executorState->es_processed = rowsProcessed;
+	execution->rowsProcessed += rowsProcessed;
 }
 
 
 /*
- * AdjustDistributedExecutionAfterLocalExecution simply updates the necessary fields of
+ * AdjustDistributedExecutionWithLocalExecution simply updates the necessary fields of
  * the distributed execution.
  */
 static void
-AdjustDistributedExecutionAfterLocalExecution(DistributedExecution *execution)
+AdjustDistributedExecutionWithLocalExecution(DistributedExecution *execution)
 {
-	/* we only need to execute the remote tasks */
-	execution->tasksToExecute = execution->remoteTaskList;
+	if (list_length(execution->localTaskList) > 0)
+	{
+		/* we only need to execute the remote tasks */
+		execution->tasksToExecute = execution->remoteTaskList;
 
-	execution->totalTaskCount = list_length(execution->remoteTaskList);
-	execution->unfinishedTaskCount = list_length(execution->remoteTaskList);
+		execution->totalTaskCount = list_length(execution->remoteTaskList);
+		execution->unfinishedTaskCount = list_length(execution->remoteTaskList);
```

Comment:
could we just assign the right values immediately in CreateDistributedExecution?

Reply:
Yes, makes sense.

To make is nicer, I did a small refactoring as well, which I wanted to do for a while now.

Basically, made the `tasksToExecute` field more explicit, and used `remoteTaskList` in the remote execution and `localTaskList` in the local execution. 

My vague memory tells me that you didn't think we need such a change, but please consider this.

Reply:
And, also realized that we should do the same change (e.g., move local execution after remote execution) for `ExecuteTaskListExtended`, that's why the refactoring seems required even more.